In [47]:
# The Shell

Let's see what's in our current directory:

In [53]:
%%bash
ls -l

total 436
-rw-rw-r--. 1 wns35789 wns35789  31785 Jan 31 23:51 BasicCommands.ipynb
-rw-rw-r--. 1 wns35789 wns35789   9235 Feb 12 16:33 Links.ipynb
-rw-rw-r--. 1 wns35789 wns35789 297484 Feb  1 23:05 LinuxCommands.ipynb
-rw-rw-r--. 1 wns35789 wns35789  13998 Feb 12 15:36 Paths.ipynb
-rw-rw-r--. 1 wns35789 wns35789  17308 Jan 31 23:51 Permissions.ipynb
-rw-rw-r--. 1 wns35789 wns35789      8 Jan 31 23:49 README.txt
drwxrwxr-x. 2 wns35789 wns35789   4096 Feb 12 17:05 shell
-rw-rw-r--. 1 wns35789 wns35789   9822 Feb 11 22:42 SpecialPermissions.ipynb
-rw-rw-r--. 1 wns35789 wns35789   5352 Feb 12 17:09 TheShell.ipynb
-rw-rw-r--. 1 wns35789 wns35789   9110 Feb 12 16:49 Untitled1.ipynb
-rw-rw-r--. 1 wns35789 wns35789     72 Feb 12 15:42 Untitled.ipynb
-rw-rw-r--. 1 wns35789 wns35789   7663 Jan 31 23:51 Wildcards.ipynb
-rwxrwxr-x. 1 wns35789 wns35789    839 Feb  1 22:33 zen2.txt
-rwxrwxrwx. 1 wns35789 wns35789    824 Feb  1 21:29 zen.txt


Let's run the same command and pass the output to the word count program 'wc':

In [54]:
%%bash
ls -l | wc

     15     128     941


'wc' says that 'ls -l' created 15 lines, 128 words and 941 characters of output.  The vertical bar (|) is called the pipe symbols and is used to pass the output of one command to the input of another.  
Linux is designed such that commands read their input from 'standard input' and write their output to what's called 'standard output".  These input and output streams are a kind of memory file (the data is never copied to disk).  Standard input is called file descriptor 0 and standard output is called file descriptor 1. 

Standard input and standard output are often called virtual files.  They are connected to the keyboard and screen by default, but can be connected elsewhere by special shell symbols.

Let's investigate these file descriptors:

In [55]:
%%bash
ls -1 1> log

Here we use file descriptor 1 (standard output) to redirect to a disk file called 'log' instead of allowing the output to go to the screen.  You can see this is so by looking at the log file:

In [56]:
%%bash
cat log

BasicCommands.ipynb
Links.ipynb
LinuxCommands.ipynb
log
Paths.ipynb
Permissions.ipynb
README.txt
shell
SpecialPermissions.ipynb
TheShell.ipynb
Untitled1.ipynb
Untitled.ipynb
Wildcards.ipynb
zen2.txt
zen.txt


File descriptor 1 is the default for output, so we could equally write:

In [57]:
%%bash
ls -l > log2
cat log

BasicCommands.ipynb
Links.ipynb
LinuxCommands.ipynb
log
Paths.ipynb
Permissions.ipynb
README.txt
shell
SpecialPermissions.ipynb
TheShell.ipynb
Untitled1.ipynb
Untitled.ipynb
Wildcards.ipynb
zen2.txt
zen.txt


Things get more interesting when a command generates errors:

In [58]:
ls *.ipynb *.unknown

ls: cannot access *.unknown: No such file or directory
BasicCommands.ipynb  Paths.ipynb               TheShell.ipynb   Wildcards.ipynb
Links.ipynb          Permissions.ipynb         Untitled1.ipynb
LinuxCommands.ipynb  SpecialPermissions.ipynb  Untitled.ipynb


Errors are sent to file descriptor 2, the standard error.  Above, we let standard output and standard error go to the screen, but we could redirect them to files:

In [26]:
ls *.ipynb *.unknown 1>log 2>errlog

Let's look at standard output:

In [27]:
%%bash
cat log

BasicCommands.ipynb
Links.ipynb
LinuxCommands.ipynb
Paths.ipynb
Permissions.ipynb
SpecialPermissions.ipynb
TheShell.ipynb
Untitled1.ipynb
Untitled.ipynb
Wildcards.ipynb


And now let's look at standard error:

In [28]:
%%bash
cat errlog

ls: cannot access *.unknown: No such file or directory


Sometimes we want standard output and error to go to the same file:

In [41]:
ls *.ipynb *.unknown 1>combinedlog 2>&1

In [42]:
%%bash
cat combinedlog

ls: cannot access *.unknown: No such file or directory
BasicCommands.ipynb
Links.ipynb
LinuxCommands.ipynb
Paths.ipynb
Permissions.ipynb
SpecialPermissions.ipynb
TheShell.ipynb
Untitled1.ipynb
Untitled.ipynb
Wildcards.ipynb


Pipes work in a similar way.  We can pass standard output and standard error to a pipe using the following rather strange notation 2>&1 (2 is sent to the same place as 1):

In [59]:
%%bash
ls *.ipynb  2>&1 | wc

     10      10     169


'ls' is not producing any errors in this case.  But modify the command a little and we get:

In [60]:
%%bash
ls *.ipynb *.unknown 2>&1 | wc

     11      19     224


Now there is an extra line corresponding to the error caused by '*.unknown'.  We can look inside the pipeline with the 'tee'command.  'tee' tees off the intermediate output to another file ('tlog'): 

In [45]:
%%bash
ls *.ipynb *.unknown 2>&1 | tee tlog | wc

     11      19     224


In [46]:
%%bash
cat tlog

ls: cannot access *.unknown: No such file or directory
BasicCommands.ipynb
Links.ipynb
LinuxCommands.ipynb
Paths.ipynb
Permissions.ipynb
SpecialPermissions.ipynb
TheShell.ipynb
Untitled1.ipynb
Untitled.ipynb
Wildcards.ipynb


Well that completes this tutorial.  Let's clean up:

In [61]:
rm *log*